pkg/internal/apis/certmanager/validation/util/nameserver.go

```diff
@@ -24,43 +24,47 @@ import (
 
 var defaultRFC2136Port = "53"
 
-// This function make a valid nameserver as per RFC2136
+// This function returns a valid nameserver (in the form <host>:<port>) for the RFC2136 provider
 func ValidNameserver(nameserver string) (string, error) {
+	nameserver = strings.TrimSpace(nameserver)
 
 	if nameserver == "" {
 		return "", fmt.Errorf("RFC2136 nameserver missing")
 	}
 
+	// When passed an IPv6, SplitHostPort expect it to be enclosed in square brackets
+	if ipaddr := net.ParseIP(nameserver); ipaddr != nil && ipaddr.To4() == nil {
+		nameserver = "[" + ipaddr.String() + "]:" + defaultRFC2136Port
+	}
+
 	// SplitHostPort Behavior
 	// nameserver          host                port    err
 	// 8.8.8.8             ""                  ""      missing port in address
 	// 8.8.8.8:            "8.8.8.8"           ""      <nil>
 	// 8.8.8.8.8:53        "8.8.8.8"           53      <nil>
+	// [2001:db8::1]       ""                  ""      missing port in address
+	// [2001:db8::1]:      "2001:db8::1"       ""      <nil>
+	// [2001:db8::1]:53    "2001:db8::1"       53      <nil>
 	// nameserver.com      ""                  ""      missing port in address
 	// nameserver.com:     "nameserver.com"    ""      <nil>
 	// nameserver.com:53   "nameserver.com"    53      <nil>
 	// :53                 ""                  53      <nil>
-	host, port, err := net.SplitHostPort(strings.TrimSpace(nameserver))
-
+	host, port, err := net.SplitHostPort(nameserver)
 	if err != nil {
 		if strings.Contains(err.Error(), "missing port") {
 			host = nameserver
+		} else {
+			return "", fmt.Errorf("RFC2136 nameserver is invalid: %s", err.Error())
 		}
 	}
 
-	if port == "" {
-		port = defaultRFC2136Port
+	if host == "" {
+		return "", fmt.Errorf("RFC2136 nameserver has no host defined, %v", nameserver)
 	}
 
-	if host != "" {
-		if ipaddr := net.ParseIP(host); ipaddr == nil {
-			return "", fmt.Errorf("RFC2136 nameserver must be a valid IP Address, not %v", host)
-		}
-	} else {
-		return "", fmt.Errorf("RFC2136 nameserver has no IP Address defined, %v", nameserver)
+	if port == "" {
+		port = defaultRFC2136Port
 	}
 
-	nameserver = host + ":" + port
-
-	return nameserver, nil
+	return net.JoinHostPort(strings.Trim(host, "[]"), port), nil
```

Comment:
Why do we trim `[]`? 🤔 can you add a comment?

Reply:
Because host might be [an enclosed IPv6 address](https://github.com/jetstack/cert-manager/pull/2682/files#diff-6f2a98d9a558374ea252c18bad645811R55) if `nameserver` is set as an enclosed IPv6 address with no port.

No problem on adding a comment (I should’ve done it), I’ll just wait on the outcome of the discussion above.

Reply:
I moved the call to Trim on line [51](https://github.com/jetstack/cert-manager/pull/2682/files#diff-6f2a98d9a558374ea252c18bad645811R51) and I added a comment.